dali/kernels/imgproc/warp_cpu.h

```diff
@@ -100,23 +101,44 @@ class WarpCPU {
     int in_w  = input.shape[1];
     int in_h  = input.shape[0];
 
-    Surface2D<OutputType> out = {
-      output.data,
-      out_w, out_h, c,
-      c, out_w*c, 1
-    };
-    Surface2D<const InputType> in = {
-      input.data,
-      in_w, in_h, c,
-      c, in_w*c, 1
-    };
+    Surface2D<const InputType> in = as_surface_channel_last(input);
 
     Sampler<static_interp, InputType> sampler(in);
 
     for (int y = 0; y < out_h; y++) {
+      OutputType *out_row = output(y, 0);
       for (int x = 0; x < out_w; x++) {
         auto src = warp::map_coords(mapping, ivec2(x, y));
-        sampler(&out(x, y), src, border);
+        sampler(&out_row[c*x], src, border);
+      }
+    }
+  }
+
+  template <DALIInterpType static_interp>
+  void RunImpl(
+      KernelContext &context,
+      const OutTensorCPU<OutputType, 3> &output,
+      const InTensorCPU<InputType, 3> &input,
+      AffineMapping<2> &mapping,
+      BorderType border = {}) {
+    // 2D HWC implementation.
+    // 3D will be added as an overload for input/output ndim == 4.
+    int out_w = output.shape[1];
+    int out_h = output.shape[0];
+    int c     = output.shape[2];
+    int in_w  = input.shape[1];
+    int in_h  = input.shape[0];
+
+    Surface2D<const InputType> in = as_surface_channel_last(input);
+
+    Sampler<static_interp, InputType> sampler(in);
+
+    vec2 dsdx = mapping.transform.col(0);
```

Comment:
Since affine mapping is linear, we can do this trick to have one vector addition instead of matrix multiplication per pixel.

Reply:
I'd write that as a comment here

Reply:
Will do. Also, I'll add tiling to prevent excessive accumulation of error in case of very wide images.

Reply:
Done.